app/bundles/EmailBundle/Tests/Helper/EmailValidatorTest.php

```diff
@@ -0,0 +1,132 @@
+
+class EmailValidatorTest extends \PHPUnit_Framework_TestCase
+{
+    public function testValidGmailEmail()
+    {
+        $this->getEmailValidator()->validate('john@gmail.com');
+    }
+
+    public function testValidGmailEmailWithPeriod()
+    {
+        $this->getEmailValidator()->validate('john.doe@gmail.com');
+    }
+
+    public function testValidGmailEmailWithPlus()
+    {
+        $this->getEmailValidator()->validate('john+doe@gmail.com');
+    }
+
+    public function testValidGmailEmailWithNonStandardTld()
+    {
+        // hopefully this domain remains intact
+        $this->getEmailValidator()->validate('john@mail.email');
+    }
+
+    public function testValidateEmailWithoutTld()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('john@doe');
+    }
+
+    public function testValidateEmailWithSpaceInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo hn@gmail.com');
+    }
+
+    public function testValidateEmailWithCaretInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo^hn@gmail.com');
+    }
+
+    public function testValidateEmailWithApostropheInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo\'hn@gmail.com');
+    }
+
+    public function testValidateEmailWithSemicolonInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo;hn@gmail.com');
+    }
+
+    public function testValidateEmailWithAmpersandInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo&hn@gmail.com');
+    }
+
+    public function testValidateEmailWithStarInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo*hn@gmail.com');
+    }
+
+    public function testValidateEmailWithPercentInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo%hn@gmail.com');
+    }
+
+    public function testValidateEmailWithDoublePeriodInIt()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('jo..hn@gmail.com');
+    }
+
+    public function testValidateEmailWithBadDNS()
+    {
+        $this->expectException(InvalidEmailException::class);
+        $this->getEmailValidator()->validate('john@doe.shouldneverexist', true);
+    }
+
+    public function testIntegrationInvalidatesEmail()
+    {
+        try {
+            $this->getEmailValidator()->doPluginValidation('bad@gmail.com');
+
+            return;
+        } catch (InvalidEmailException $exception) {
+            if ('bad email' === $exception->getMessage()) {
+                return;
```

Comment:
Why this function has 2 return statements? E-mail should either pass (1st return) or throw an exception (2nd return).
I can see nothing bad in email `bad@gmail.com`.

Reply:
Right; first return isn't required so moved it. The bad@gmail.com is just a placeholder that's caught by the fake listener class to test integrations https://github.com/mautic/mautic/blob/6f578d5a1f5cea03c1bcbfb6efe3a835b5704cb4/app/bundles/EmailBundle/Tests/Helper/EventListener/EmailValidationSubscriber.php#L39.